src/main/kotlin/com/terraformation/backend/gis/geoserver/GeoServerClient.kt

```diff
@@ -0,0 +1,100 @@
+package com.terraformation.backend.gis.geoserver
+
+import com.fasterxml.jackson.databind.DeserializationFeature
+import com.fasterxml.jackson.databind.MapperFeature
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.dataformat.xml.XmlMapper
+import com.fasterxml.jackson.module.kotlin.registerKotlinModule
+import com.terraformation.backend.config.TerrawareServerConfig
+import io.ktor.client.HttpClient
+import io.ktor.client.call.body
+import io.ktor.client.engine.java.Java
+import io.ktor.client.plugins.auth.Auth
+import io.ktor.client.plugins.auth.providers.BasicAuthCredentials
+import io.ktor.client.plugins.auth.providers.basic
+import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
+import io.ktor.client.plugins.defaultRequest
+import io.ktor.client.request.get
+import io.ktor.client.request.parameter
+import io.ktor.client.request.request
+import io.ktor.client.statement.HttpResponse
+import io.ktor.http.ContentType
+import io.ktor.http.HttpMethod
+import io.ktor.serialization.jackson.JacksonConverter
+import jakarta.inject.Named
+import jakarta.ws.rs.core.MediaType
+import kotlinx.coroutines.runBlocking
+
+@Named
+class GeoServerClient(
+    private val config: TerrawareServerConfig,
+    private val objectMapper: ObjectMapper,
+) {
+  private val wfsVersion = "2.0.0"
+  private val httpClient: HttpClient by lazy { createHttpClient() }
+
+  fun getCapabilities(): WfsCapabilities {
+    return sendGetRequest("GetCapabilities")
+  }
+
+  private inline fun <reified T> sendGetRequest(
+      command: String,
+      params: Map<String, Any> = emptyMap()
+  ): T {
+    return runBlocking { sendRequest(HttpMethod.Get, command, params).body() }
+  }
+
+  private suspend fun sendRequest(
+      requestMethod: HttpMethod,
+      command: String,
+      params: Map<String, Any> = emptyMap(),
+  ): HttpResponse {
+    return httpClient.request {
+      method = requestMethod
+      parameter("request", command)
+      parameter("service", "wfs")
+      parameter("version", wfsVersion)
+      params.forEach { parameter(it.key, it.value.toString()) }
+
+      // Prefer JSON if supported for the operation; some operations only support XML and
+      // will ignore this parameter.
+      parameter("outputFormat", MediaType.APPLICATION_JSON)
+    }
+  }
+
+  private fun createHttpClient(): HttpClient {
+    val wfsUrl =
+        config.geoServer.wfsUrl ?: throw IllegalStateException("GeoServer URL not configured")
```

Comment:
Not sure if it matters or not, but is there any concern about the url not being configured and us not knowing about it until someone makes a geoserver request because of the lazy instantiation here?

Reply:
The idea is to not require people to set this up in dev environments where they aren't interacting with GeoServer. We use the same pattern for our other external integrations like HubSpot and Atlassian.